Show badge for member only videos (Local API)

PikachuEXE opened the pull request:
<!-- Thanks for sending a pull request! Make sure to follow the contributing guidelines. -->
<!-- Important note, we may remove your pull request if you do not use this provided PR template correctly. -->

## Pull Request Type
<!-- Please select what type of pull request this is: [x] -->
- [ ] Bugfix
- [x] Feature Implementation
- [ ] Documentation
- [ ] Other

## Related issue
<!-- Please link the issue your pull request is referring to. -->
<!-- If this pull request fully resolves the relevant issue, put "closes" before the issue number. -->
<!-- Example: "closes #123456". -->
Related to https://github.com/FreeTubeApp/FreeTube/issues/6425 but that issue is for hiding
This one is for showing a badge only (no hiding)

## Description
<!-- Please write a clear and concise description of what the pull request does. -->
There are member only videos in subscription/channel video list
This PR add a badge for them (actually replaces the duration display
IV API has no such attribute so can't implement for it

## Screenshots <!-- If appropriate -->
<!-- Please add before and after screenshots if there is a visible change. -->
![image](https://github.com/user-attachments/assets/3f5df7c2-28a1-4ed6-b19b-b2f6e574cf34)
![image](https://github.com/user-attachments/assets/a7409639-21a2-4cd3-aea9-6aa2d26b4fc8)


## Testing <!-- for code that is not small enough to be easily understandable -->
<!-- Has this pull request been tested? -->
<!-- Please describe shortly how you tested it. -->
<!-- Are there any ramifications remaining? -->
- Visit channel(s) with member only videos (e.g. https://www.youtube.com/channel/UCXuqSBlHAE6Xw-yeJA0Tunw
- Ensure badge shown

## Desktop
<!-- Please complete the following information-->
- **OS:**
- **OS Version:**
- **FreeTube version:**

## Additional context
<!-- Add any other context about the pull request here. -->

absidue commented:
My understanding was that we were going to always hide them from the channel and subscription pages, by skipping them during the parsing, like we do with the members-only section on the channel home tab, without a setting or labelling them. There is no point showing videos that you have no way of watching in FreeTube (you require a Google account that is paying a subscription to the channel in question).

So that begs the question, is this just a temporary PR to learn how to identify members-only videos and you'll revert the changes and implement the skipping in a follow up or have you decided to change plans?

PikachuEXE commented:
I think it depends on the channel
I see some channels just posting members only videos as an early release to supporters. 
For those I am not sure if users want to use bookmark them early or whatever.
Also IV API is not supporting members only property yet so showing a badge can at least be a stop gap solution until IV API also supports it. (i.e. Local API provides more info and we just show it instead of just hiding some videos in local API which might seem a bug

I might just hide them if IV API supports it already

efb4f5ff-1298-471a-8973-3d47447115dc commented:
The plan is definitely to go back to state FT was in before. Dont show member only content anywhere like explained in https://github.com/FreeTubeApp/FreeTube/pull/6437

At this point we have many features that are only on Local API that arent available on IV. 

I think the route needs to be:
- Create PR here to filter on Local API
- Create PR on IV side to make filtering available to us

PikachuEXE commented:
I will wait for https://github.com/FreeTubeApp/FreeTube/pull/7187 to avoid merge conflict...

PikachuEXE commented:
Updated to hide member only videos
Don't have example for hash tag view

PikachuEXE commented:
I got videos with `Members first` (e.g. https://www.youtube.com/channel/UCy8E4XShrcph5WWLg05zETQ
Hide?
![2025-04-19 15_43_44-Window](https://github.com/user-attachments/assets/d3084f9a-e47a-4c55-b0f4-96415a146884)

efb4f5ff-1298-471a-8973-3d47447115dc commented:
Yes, hide it. Nice catch.

How it looks like on YT

![firefox_ti48RI89vf](https://github.com/user-attachments/assets/68a602aa-6fc0-46f1-9c9d-4b5221f2e5bb)

PikachuEXE commented:
> I wonder how it will present it in FT when it becomes available to everyone

Just normal video

Also I never see member only/first videos in home tab

absidue commented:
The channel home tab has a dedicated member's only section and we already skip parsing that entire section (unlike the approach in this PR which parses members-only stuff and then filters it out right at the end), so I'm pretty sure they won't show any other members-only stuff there. The only place I've seen members-only and members-first is on the channel videos tab, which is is also used on the subscriptions page.

efb4f5ff-1298-471a-8973-3d47447115dc commented:
FYI: PR on IV side https://github.com/iv-org/invidious/pull/5222. Do we need to do something after that PR has merged?

PikachuEXE commented:
Yes and that PR does not include member first currently, already commented there

PikachuEXE commented:
Added Member first handling but not tested yet, see updated testing section
Gonna test it myself with custom build

PikachuEXE commented:
14:50 UTC+8
![2025-04-20 14_47_50-Window](https://github.com/user-attachments/assets/270e2a8c-9cae-4aaa-9c7a-061ea6470dc7)
![2025-04-20 14_49_41-Window](https://github.com/user-attachments/assets/828da2c3-ab09-4b70-88d2-a943e55e6067)
![2025-04-20 14_51_01-Window](https://github.com/user-attachments/assets/f2044714-f185-4d07-9951-c2905f0385bc)

absidue reviewed (commented):
Instead of adding extra properties and creating extra arrays all over FreeTube for something that only appears on the channel videos tab (and therefore also on the subscription videos tab), to me at least it makes a lot more sense to do the filtering in the `parseLocalChannelVideos` function in the `src/renderer/helpers/api/local.js`. Something like this (`BADGE_STYLE_TYPE_MEMBERS_ONLY` is used for both members only and members first videos):
```diff
export function parseLocalChannelVideos(videos, channelId, channelName) {
-  return videos.map((video) => parseLocalListVideo(video, channelId, channelName))
+  const parsedVideos = []
+
+  for (const video of videos) {
+    if (!video.is(YTNodes.Video) || !video.badges.some(badge => badge.style === 'BADGE_STYLE_TYPE_MEMBERS_ONLY')) {
+      parsedVideos.push(parseLocalListVideo(video, channelId, channelName))
+    }
  }

  return parsedVideos
}
```

PikachuEXE commented:
Updated but the member first channel example doesn't have video a.t.m. so I can't test yet...

PikachuEXE commented:
Anyone got more example channels with member first videos...?

efb4f5ff-1298-471a-8973-3d47447115dc commented:
> Anyone got more example channels with member first videos...?

Nope couldnt find any but i trust the code

efb4f5ff-1298-471a-8973-3d47447115dc commented:
Member first channel has a video but i just didnt notice it

![VirtualBoxVM_9dTy2OpLxX](https://github.com/user-attachments/assets/2c6774ce-2c4b-41ad-abb1-46053d79043a)
![firefox_iQX5RzTNJ2](https://github.com/user-attachments/assets/5a633af4-fc8a-4c7a-83ca-134602fdeda9)

PikachuEXE commented:
I got a bug in video searching in channel, fixing today

absidue commented:
Probably because you left all the other code in when you added in my suggestion, instead of using it as a replacement as it was intended to be.

PikachuEXE commented:
Updated, channel searching code was updated to use `forEach` but I forgot to remove `filter` after it
Fixed that and also now using `reduce` instead of declaring an array then `forEach`
Tested channel searching and haven't tested Hashtag again due to lack of example

Member first re-tested like https://github.com/FreeTubeApp/FreeTube/pull/7208#issuecomment-2831363667

absidue commented on src/renderer/helpers/api/local.js:
```diff
@@ -1180,7 +1190,9 @@ export function parseLocalListVideo(item, channelId, channelName) {
       isVr180: video.badges.some(badge => badge.label === 'VR180'),
       isVr360: video.badges.some(badge => badge.label === '360°'),
       is3d: video.badges.some(badge => badge.label === '3D'),
-      hasCaptions: video.has_captions
+      isMemberOnly: video.badges.some(badge => badge.label === 'Members only'),
+      isMemberFirst: video.badges.some(badge => badge.label === 'Members first'),
+      hasCaptions: video.has_captions,
```
Revert

absidue commented on src/renderer/helpers/subscriptions.js:
```diff
@@ -64,6 +64,10 @@ export function updateVideoListAfterProcessing(videos) {
     })
   }
 
+  videoList = videoList.filter(item => {
+    return !item.isMemberOnly && !item.isMemberFirst
+  })
+
```
Ditto

absidue commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -1143,7 +1143,9 @@ async function getChannelVideosLocalMore() {
       const continuation = await videoContinuationData.value.getContinuation()
 
-      latestVideos.value = latestVideos.value.concat(parseLocalChannelVideos(continuation.videos, id.value, channelName.value))
+      latestVideos.value = latestVideos.value.concat(
+        parseLocalChannelVideos(continuation.videos, id.value, channelName.value)
+      )
```
Ditto

absidue commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -1420,7 +1422,9 @@ async function getChannelLiveLocalMore() {
      */
     const continuation = await liveContinuationData.value.getContinuation()
 
-    latestLive.value = latestLive.value.concat(parseLocalChannelVideos(continuation.videos, id.value, channelName.value))
+    latestLive.value = latestLive.value.concat(
+      parseLocalChannelVideos(continuation.videos, id.value, channelName.value)
+    )
```
Ditto

absidue commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -2047,14 +2051,17 @@ async function searchChannelLocal() {
       .flatMap(itemSection => itemSection.contents)
-      .filter(item => item.type === 'Video' || (!hideChannelPlaylists.value && item.type === 'Playlist'))
-      .map(item => {
+      .reduce((items, item) => {
         if (item.type === 'Video') {
-          return parseLocalListVideo(item)
-        } else {
-          return parseLocalListPlaylist(item, id.value, channelName.value)
+          const video = parseLocalListVideo(item)
+          if (video.isMemberOnly || video.isMemberFirst) { return null }
+
+          items.push(video)
+        } else if (!hideChannelPlaylists.value && item.type === 'Playlist') {
+          items.push(parseLocalListPlaylist(item, id.value, channelName.value))
         }
-      })
+        return items
+      }, [])
```
Ditto

absidue commented on src/renderer/views/Hashtag/Hashtag.vue:
```diff
@@ -148,7 +148,13 @@ async function getInvidiousHashtag(page = 1) {
   try {
     const hashtagData = await getHashtagLocal(hashtag.value)
-    videos.value = hashtagData.videos.map((video) => parseLocalListVideo(video))
+    videos.value = hashtagData.videos.reduce((results, video) => {
+      const v = parseLocalListVideo(video)
+      if (v.isMemberOnly || v.isMemberFirst) {
+        results.push(v)
+      }
+      return results
+    }, [])
```
Ditto

absidue commented on src/renderer/views/Hashtag/Hashtag.vue:
```diff
@@ -171,7 +177,13 @@ async function getLocalHashtag() {
 async function getLocalHashtagMore() {
   try {
     const continuation = await hashtagContinuationData.value.getContinuation()
-    const newVideos = continuation.videos.map((video) => parseLocalListVideo(video))
+    const newVideos = continuation.videos.reduce((results, video) => {
+      const v = parseLocalListVideo(video)
+      if (v.isMemberOnly || v.isMemberFirst) {
+        results.push(v)
+      }
+      return results
+    }, [])
```
Ditto

absidue commented on src/renderer/views/Watch/Watch.js:
```diff
@@ -1712,7 +1712,8 @@ export default defineComponent({
       return channelsHidden.some(ch => ch.name === video.authorId) ||
         channelsHidden.some(ch => ch.name === video.author) ||
-        forbiddenTitles.some((text) => video.title?.toLowerCase().includes(text.toLowerCase()))
+        forbiddenTitles.some((text) => video.title?.toLowerCase().includes(text.toLowerCase())) ||
+        (video.isMemberOnly || video.isMemberFirst)
```
Ditto

absidue reviewed (commented):
> Instead of adding extra properties and creating extra arrays all over FreeTube for something that only appears on the channel videos tab (and therefore also on the subscription videos tab), to me at least it makes a lot more sense to do the filtering in the `parseLocalChannelVideos` function in the `src/renderer/helpers/api/local.js`

This is still not properly addressed, yes you copied my provided code snippet but didn't actually remove any of the other code. I know you might not care about performance but I do, so I would feel uncomfortable  approving and merging something that adds filtering all over FreeTube when it is only needed in one single place. I'm using just that code snippet in my custom build and it is working great. We could always merge this as is and then I could do a follow up PR to clean it up, but that seems wrong too, so it is probably better to get it right in this pull request.

If you think about it the fact that member's only videos only appear on the channel page makes a lot of sense, there it's a "if you liked other videos by this channel you may also like their members only videos" but on the hashtag page, nobody is going to decide to sign up for a subscription for a channel they've not watched any videos from, just to watch one members only video.

PikachuEXE commented:
Updated, was mostly trying to quick fix the bug I introduce so I can get a custom build that works
And I did not mean that it's all perfect and ready for merge

I know I suck in many ways, performance side included. So I have no problem receiving criticism on implementations and act if necessary.
But I am not comfortable hearing `I know you might not care about performance but I do`. This kind of speculation should never be said to any current/wanna be contributor.
It's why we have code review.

efb4f5ff-1298-471a-8973-3d47447115dc commented:
> But I am not comfortable hearing `I know you might not care about performance but I do`. This kind of speculation should never be said to any current/wanna be contributor.
> It's why we have code review. 

This was indeed not necessary to say. A review can be critical but shouldnt be personal.

 > I know I suck in many ways

Didnt notice any of them but if you do then you also dont suck in many ways :)

JamesClarke7283 commented:
Ive been using this PR and its been working fine, i haven't been doing performance testing though.

Do we have a known workflow for performance testing here? if someone has one they use it might be good to share and refer to later. I couldn't find anything on testing in the `CONTRIBUTING.md` or wiki.

Using at this commit:
 `https://github.com/PikachuEXE/FreeTube/commit/fd800c`

absidue reviewed (approved):
I'm really sorry for how I expressed myself, you definitely didn't deserve those harsh words.

PikachuEXE commented:
@ChunkyProgrammer your power is needed
![image](https://github.com/user-attachments/assets/6d91e32d-485f-40d8-97be-2f8d09230791)
